pkg/syncer/mutators/deployment.go

```diff
@@ -0,0 +1,247 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	utilspointer "k8s.io/utils/pointer"
+)
+
+type DeploymentMutator struct {
+	fromConfig *rest.Config
+	kcpHost    string
+}
+
+func (dt *DeploymentMutator) getGVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "deployments",
+	}
+}
+
+func (dt *DeploymentMutator) Register(mutators map[schema.GroupVersionResource]Mutator) {
+	if _, ok := mutators[dt.getGVR()]; !ok {
+		mutators[dt.getGVR()] = dt
+	}
+}
+
+func NewDeploymentMutator(fromConfig *rest.Config, kcpHost string) *DeploymentMutator {
+	return &DeploymentMutator{
+		fromConfig: fromConfig,
+		kcpHost:    kcpHost,
+	}
+}
+
+// ApplySpec makes modifications to the Spec of the deployment object.
+func (dt *DeploymentMutator) Mutate(downstreamObj *unstructured.Unstructured) error {
+	var deployment appsv1.Deployment
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		downstreamObj.UnstructuredContent(),
+		&deployment)
+	if err != nil {
+		return err
+	}
+
+	// If the deployment has not serviceaccount defined or it is "default", that means that it is using the default one
+	// and we need will need to override it with our own, "kcp-default"
+	//
+	// If the deployment has a serviceaccount defined, that means we need to synchronize the created service account
+	// down to the workloadcluster, so we don't modify it, and expect the scheduler to do the job.
+	//
+	if deployment.Spec.Template.Spec.ServiceAccountName == "" || deployment.Spec.Template.Spec.ServiceAccountName == "default" {
+		deployment.Spec.Template.Spec.ServiceAccountName = "kcp-default"
+	}
+
+	// Setting AutomountServiceAccountToken to false allow us to control the ServiceAccount
+	// VolumeMount and Volume definitions.
+	deployment.Spec.Template.Spec.AutomountServiceAccountToken = utilspointer.BoolPtr(false)
+
+	var hostname, port string
+	if dt.kcpHost != "" {
+		hostname, port, err = net.SplitHostPort(dt.kcpHost)
+		if err != nil {
+			return err
+		}
+	} else {
+		u, err := url.Parse(dt.fromConfig.Host)
+		if err != nil {
+			return err
+		}
+		hostname = u.Hostname()
+		port = u.Port()
+	}
+
+	kcpExternalPort := port
+	kcpExternalHost := hostname
+
+	overrideEnvs := []corev1.EnvVar{
+		{Name: "KUBERNETES_SERVICE_PORT", Value: kcpExternalPort},
+		{Name: "KUBERNETES_SERVICE_PORT_HTTPS", Value: kcpExternalPort},
+		{Name: "KUBERNETES_SERVICE_HOST", Value: kcpExternalHost},
+	}
+
+	// This is the VolumeMount that we will append to all the containers of the deployment
+	serviceAccountMount := corev1.VolumeMount{
+		Name:      "kcp-api-access",
+		MountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
+		ReadOnly:  true,
+	}
+
+	// This is the Volume that we will add to the Deployment in order to control
+	// the name of the ca.crt references (kcp-root-ca.crt vs kube-root-ca.crt)
```

Comment:
Maybe add a bit more detail as to why this renaming is necessary (e.g. kube already writes its own ca secret so we need to differentiate the one for kcp)?